File: src/relax/frontend/nn/visitor.cc
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

/*!
 * \file src/relax/frontend/nn/visitor.cc
 * \brief C++ port of python/tvm/relax/frontend/nn/visitor.py
 *
 * Implements the Mutator class: a visitor/mutator for nn::Module trees.
 *
 * Design notes
 * ------------
 * The Python Mutator walks a Module's __dict__ to find children.  In C++
 * the equivalent is the NNModuleNode::attrs map, which is populated by
 * every Make* factory function via PopulateAttrs.  ModuleList and
 * ModuleDict are handled as special cases (they store children in their
 * own `modules` field rather than `attrs`).
 *
 * Dispatch priority (matches Python visitor.py):
 *   ModuleDictNode  -> visit_moduledict
 *   ModuleListNode  -> visit_modulelist
 *   EffectNode      -> visit_effect
 *   ParameterNode   -> visit_param
 *   NNModuleNode    -> visit_module
 *
 * All visit_* defaults delegate to visit(), which recurses into children.
 * visit() returns the (possibly replaced) node as ffi::Any so that
 * subclasses can substitute a different object at any level.
 *
 * Python–C++ correspondence
 * -------------------------
 *   Python                     C++
 *   ──────────────────────     ──────────────────────────────────────
 *   isinstance(n, ModuleDict)  n.as<ModuleDictNode>()
 *   isinstance(n, ModuleList)  n.as<ModuleListNode>()
 *   isinstance(n, Effect)      n.as<EffectNode>()
 *   isinstance(n, Parameter)   n.as<ParameterNode>()
 *   isinstance(n, Module)      n.as<NNModuleNode>()
 *   node.__dict__.items()      mod->attrs (ffi::Map<String,Any>)
 *   node[i] = ...              list->modules.Set(i, ...)
 *   node[k] = ...              dict->modules.Set(k, ...)
 *   setattr(node, k, v)        mod->attrs.Set(k, v)
 */

#include "visitor.h"

#include <string>

namespace tvm {
namespace relax {
namespace frontend {
namespace nn {

// ---------------------------------------------------------------------------
// Internal helpers
// ---------------------------------------------------------------------------

/*static*/ std::string Mutator::ChildName(const std::string& parent, const std::string& child) {
  // Mirrors _get_child_name() in visitor.py:
  //   if parent == "": return child
  //   else:            return f"{parent}.{child}"
  if (parent.empty()) return child;
  return parent + "." + child;
}

ffi::Any Mutator::DispatchChild(const std::string& name, ffi::Any value) {
  // Try to cast to ObjectRef; non-object values (scalars, strings, etc.)
  // are returned unchanged — they are not module children.
  auto opt_ref = value.try_cast<runtime::ObjectRef>();
  if (!opt_ref.has_value() || !opt_ref.value().defined()) return value;
  runtime::ObjectRef obj = opt_ref.value();

  // Dispatch in the same priority order as visitor.py.
  if (obj->IsInstance<ModuleDictNode>()) {
    return visit_moduledict(name, Downcast<ModuleDict>(obj));
  }
  if (obj->IsInstance<ModuleListNode>()) {
    return visit_modulelist(name, Downcast<ModuleList>(obj));
  }
  if (obj->IsInstance<EffectNode>()) {
    return visit_effect(name, obj);
  }
  if (obj->IsInstance<ParameterNode>()) {
    return visit_param(name, Downcast<NNParameter>(obj));
  }
  if (obj->IsInstance<NNModuleNode>()) {
    return visit_module(name, obj);
  }
  // Not a recognised module child — return unchanged.
  return value;
}

// ---------------------------------------------------------------------------
// visit() — the main traversal driver
// ---------------------------------------------------------------------------

ffi::Any Mutator::visit(const std::string& name, ffi::Any node) {
  auto opt_ref = node.try_cast<runtime::ObjectRef>();
  if (!opt_ref.has_value() || !opt_ref.value().defined()) return node;
  runtime::ObjectRef obj = opt_ref.value();

  // ── ModuleList ────────────────────────────────────────────────────────────
  // Python:
  //   for i in range(len(node)):
  //       node[i] = dispatch(f"{name}.{i}", node[i])
  if (const auto* list_node = obj.as<ModuleListNode>()) {
    ModuleList ml = Downcast<ModuleList>(obj);
    ffi::Array<ffi::Any> new_modules = list_node->modules;
    for (int64_t i = 0; i < static_cast<int64_t>(new_modules.size()); ++i) {
      std::string child_name = ChildName(name, std::to_string(i));
      ffi::Any old_val = new_modules[i];
      ffi::Any new_val = DispatchChild(child_name, old_val);
      if (!new_val.same_as(old_val)) {
        new_modules.Set(i, new_val);
      }
    }
    ml->modules = new_modules;
    return ffi::Any(ml);
  }

  // ── ModuleDict ────────────────────────────────────────────────────────────
  // Python:
  //   for k, v in node.items():
  //       node[k] = dispatch(_get_child_name(name, k), v)
  if (const auto* dict_node = obj.as<ModuleDictNode>()) {
    ModuleDict md = Downcast<ModuleDict>(obj);
    ffi::Map<ffi::String, ffi::Any> new_modules = dict_node->modules;
    for (const auto& [k, v] : dict_node->modules) {
      std::string child_name = ChildName(name, std::string(k));
      ffi::Any new_val = DispatchChild(child_name, v);
      if (!new_val.same_as(v)) {
        new_modules.Set(k, new_val);
      }
    }
    md->modules = new_modules;
    return ffi::Any(md);
  }

  // ── Module (generic NNModuleNode) ─────────────────────────────────────────
  // Python:
  //   for key, value in node.__dict__.items():
  //       setattr(node, key, dispatch(_get_child_name(name, key), value))
  if (const auto* mod_node = obj.as<NNModuleNode>()) {
    // Work on a copy of the attrs map so we can mutate it safely.
    ffi::Map<ffi::String, ffi::Any> new_attrs = mod_node->attrs;
    for (const auto& [k, v] : mod_node->attrs) {
      std::string child_name = ChildName(name, std::string(k));
      ffi::Any new_val = DispatchChild(child_name, v);
      if (!new_val.same_as(v)) {
        new_attrs.Set(k, new_val);
      }
    }
    // NNModuleNode::_type_mutable = true, so operator->() returns NNModuleNode*.
    const_cast<NNModuleNode*>(mod_node)->attrs = new_attrs;
    return node;
  }

  // Not a module node — return unchanged.
  return node;
}

// ---------------------------------------------------------------------------
// Default visit_* implementations — all delegate back to visit()
// ---------------------------------------------------------------------------

ffi::Any Mutator::visit_module(const std::string& name, runtime::ObjectRef node) {
  return visit(name, ffi::Any(node));
}

ffi::Any Mutator::visit_effect(const std::string& name, runtime::ObjectRef node) {
  return visit(name, ffi::Any(node));
}

ffi::Any Mutator::visit_param(const std::string& name, NNParameter node) {
  return visit(name, ffi::Any(node));
}

ffi::Any Mutator::visit_moduledict(const std::string& name, ModuleDict node) {
  return visit(name, ffi::Any(node));
}

ffi::Any Mutator::visit_modulelist(const std::string& name, ModuleList node) {
  return visit(name, ffi::Any(node));
}

}  // namespace nn
}  // namespace frontend
}  // namespace relax
}  // namespace tvm

File: python/tvm/relax/frontend/nn/visitor.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.
"""The visitor and mutator infra for nn.Module."""

from typing import Any

from . import core as nn


class Mutator:
    """The mutator for nn.Module transform. Users can override the ``visit_*``
    methods to apply transforms in different structures, or override ``visit``
    to change the traversal strategy entirely."""

    def visit_module(self, name: str, node: nn.Module) -> Any:
        """Visit an nn.Module node.

        Parameters
        ----------
        name : str
            Dotted path of this node within its parent.
        node : nn.Module
            The module node to visit.

        Returns
        -------
        ret_node : Any
            The (possibly replaced) node.
        """
        return self.visit(name, node)

    def visit_effect(self, name: str, node: nn.Effect) -> Any:
        """Visit an nn.Effect node.

        Parameters
        ----------
        name : str
            Dotted path of this node within its parent.
        node : nn.Effect
            The effect node to visit.

        Returns
        -------
        ret_node : Any
            The (possibly replaced) node.
        """
        return self.visit(name, node)

    def visit_param(self, name: str, node: nn.Parameter) -> Any:
        """Visit an nn.Parameter node.

        Parameters
        ----------
        name : str
            Dotted path of this node within its parent.
        node : nn.Parameter
            The parameter to visit.

        Returns
        -------
        ret_node : Any
            The (possibly replaced) parameter.
        """
        return self.visit(name, node)

    def visit_moduledict(self, name: str, node: nn.ModuleDict) -> Any:
        """Visit an nn.ModuleDict node.

        Parameters
        ----------
        name : str
            Dotted path of this node within its parent.
        node : nn.ModuleDict
            The ModuleDict to visit.

        Returns
        -------
        ret_node : Any
            The (possibly replaced) node.
        """
        return self.visit(name, node)

    def visit_modulelist(self, name: str, node: nn.ModuleList) -> Any:
        """Visit an nn.ModuleList node.

        Parameters
        ----------
        name : str
            Dotted path of this node within its parent.
        node : nn.ModuleList
            The ModuleList to visit.

        Returns
        -------
        ret_node : Any
            The (possibly replaced) node.
        """
        return self.visit(name, node)

    def visit(self, name: str, node: Any) -> Any:
        """Dispatch driver: recurse into the module tree and call the
        appropriate ``visit_*`` method for each child.

        Parameters
        ----------
        name : str
            Dotted path of *node* within its parent (pass ``""`` at root).
        node : Any
            The node to visit.

        Returns
        -------
        ret_node : Any
            The (possibly mutated) node.
        """

        def _get_child_name(parent: str, child: str) -> str:
            if parent == "":
                return child
            return f"{parent}.{child}"

        _MODULEDICT_KEY = "relax.frontend.nn.ModuleDict"
        _MODULELIST_KEY = "relax.frontend.nn.ModuleList"

        def _type_key(value: Any) -> str | None:
            get_key = getattr(value, "GetTypeKey", None)
            if get_key is not None:
                return get_key()
            return None

        def _dispatch(child_name: str, value: Any) -> Any:
            key = _type_key(value)
            if key == _MODULEDICT_KEY or isinstance(value, nn.ModuleDict):
                return self.visit_moduledict(child_name, value)
            if key == _MODULELIST_KEY or isinstance(value, nn.ModuleList):
                return self.visit_modulelist(child_name, value)
            if isinstance(value, nn.Effect):
                return self.visit_effect(child_name, value)
            if isinstance(value, nn.Parameter):
                return self.visit_param(child_name, value)
            if isinstance(value, nn.Module):
                return self.visit_module(child_name, value)
            return value

        if isinstance(node, nn.ModuleList):
            for i in range(len(node)):
                child_name = _get_child_name(name, str(i))
                old = node[i]
                new = _dispatch(child_name, old)
                if new is not old:
                    node[i] = new
        elif isinstance(node, nn.ModuleDict):
            for k, v in node.items():
                child_name = _get_child_name(name, k)
                new = _dispatch(child_name, v)
                if new is not v:
                    node[k] = new
        elif isinstance(node, nn.Module):
            for key, value in node.__dict__.items():
                child_name = _get_child_name(name, key)
                new = _dispatch(child_name, value)
                if new is not value:
                    setattr(node, key, new)
        # Non-module values (int, str, etc.) are returned unchanged.
        return node
